src/hazelcore/semantics/UHPat.re

```diff
@@ -153,6 +156,35 @@ and make_inconsistent_operand =
     (Parenthesized(set_p), u_gen);
   };
 
+let patterns_of_type =
+    (u_gen: MetaVarGen.t, ty: HTyp.t): (list(t), MetaVarGen.t) => {
+  let (hole1, u_gen) = u_gen |> new_EmptyHole;
+  let (hole2, u_gen) = u_gen |> new_EmptyHole;
+  let (hole3, u_gen) = u_gen |> new_EmptyHole;
+  let (hole4, u_gen) = u_gen |> new_EmptyHole;
+  let pattern =
+    OpSeq.(
+      switch (ty) {
+      | HTyp.Hole
+      | Unit
+      | Arrow(_, _) => []
+      | Num => [wrap(hole1), wrap(hole2)]
+      | Bool => List.map(b => wrap(boollit(b)), [true, false])
+      | Prod(_, _) => [
```

Comment:
for products, you'll want to figure out how many commas to put in -- you can ask Qiran for pointers to his code to do this...

Reply:
Alright, I'll have a look